File: src/sys/java/fan/sys/StrBuf.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   28 Dec 05  Brian Frank  Creation
//
package fan.sys;

import fanx.main.*;

/**
 * StrBuf mutable random-access sequence of integer characters.
 */
public class StrBuf
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  /**
   * Create with initial capacity of 16.
   */
  public static StrBuf make()
  {
    return new StrBuf(new StringBuilder(16));
  }

  /**
   * Create with specified capacity.
   */
  public static StrBuf make(long capacity)
  {
    return new StrBuf(new StringBuilder((int)capacity));
  }

  public StrBuf(StringBuilder sb)
  {
    this.sb = sb;
  }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  public boolean isEmpty()
  {
    return sb.length() == 0;
  }

  public long size()
  {
    return sb.length();
  }

  public long capacity()
  {
    return sb.capacity();
  }

  public void capacity(long size)
  {
    sb.ensureCapacity((int)size);
  }

  public long get(long index)
  {
    int i = (int)index;
    if (i < 0) i = sb.length()+i;
    return sb.charAt(i);
  }

  public String getRange(Range r)
  {
    int size = sb.length();

    int s = (int)r.startIndex(size);
    int e = (int)r.endIndex(size);
    if (e+1 < s) throw IndexErr.make(""+r);

    return sb.substring(s, e+1);
  }

  public StrBuf set(long index, long ch)
  {
    int i = (int)index;
    if (i < 0) i = sb.length()+i;
    sb.setCharAt(i, (char)ch);
    return this;
  }

  public StrBuf add(Object x)
  {
    String s = (x == null) ? "null" : toStr(x);
    sb.append(s);
    return this;
  }

  public StrBuf addChar(long ch)
  {
//  sb.append((char)ch);
	sb.appendCodePoint((int)ch);
    return this;
  }
  
  public StrBuf addStr(String str, long off, long len)
  {
	sb.append(str, (int)off, (int)(off+len));
    return this;
  }

  public StrBuf join(Object x) { return join(x, " "); }
  public StrBuf join(Object x, String sep)
  {
    String s = (x == null) ? "null" : toStr(x);
    if (sb.length() > 0) sb.append(sep);
    sb.append(s);
    return this;
  }

  public StrBuf insert(long index, Object x)
  {
    String s = (x == null) ? "null" : toStr(x);
    int i = (int)index;
    if (i < 0) i = sb.length()+i;
    if (i > sb.length()) throw IndexErr.make(""+index);
    sb.insert(i, s);
    return this;
  }

  public StrBuf remove(long index)
  {
    int i = (int)index;
    if (i < 0) i = sb.length()+i;
    if (i >= sb.length()) throw IndexErr.make(""+index);
    sb.delete(i, i+1);
    return this;
  }

  public StrBuf removeRange(Range r)
  {
    int s = (int)r.startIndex(sb.length());
    int e = (int)r.endIndex(sb.length());
    int n = e - s + 1;
    if (n < 0) throw IndexErr.make(""+r);
    sb.delete(s, e+1);
    return this;
  }

  public StrBuf replaceRange(Range r, String str)
  {
    int s = (int)r.startIndex(sb.length());
    int e = (int)r.endIndex(sb.length());
    if (e+1 < s) throw IndexErr.make(""+r);
    sb.replace(s, e+1, str);
    return this;
  }

  public StrBuf clear()
  {
    sb.setLength(0);
    return this;
  }

  public String toStr()
  {
    return sb.toString();
  }

  private static Type type;
  public Type typeof()
  {
	if (type == null) type = Sys.findType("sys::StrBuf");
    return type;
  }
//
//  public OutStream out()
//  {
//    return new StrBufOutStream(this);
//  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  public StringBuilder sb;

}

File: src/sys/js/StrBuf.js
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   29 Dec 08  Andy Frank  Creation
//   20 May 09  Andy Frank  Refactor to new OO model
//

/**
 * StrBuf
 */
fan.sys.StrBuf = fan.sys.Obj.$extend(fan.sys.Obj);

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

fan.sys.StrBuf.prototype.$ctor = function()
{
  this.m_str = "";
}

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

fan.sys.StrBuf.prototype.$typeof = function()
{
  return fan.sys.StrBuf.$type;
}

fan.sys.StrBuf.prototype.add = function(obj)
{
  this.m_str += obj==null ? "null" : fan.sys.ObjUtil.toStr(obj);
  return this;
}

fan.sys.StrBuf.prototype.addChar = function(ch)
{
  this.m_str += String.fromCharCode(ch);
  return this;
}

fan.sys.StrBuf.prototype.capacity = function()
{
  if (this.m_capacity == null) return this.m_str.length;
  return this.m_capacity;
}
fan.sys.StrBuf.prototype.capacity$ = function(c) { this.m_capacity = c; }
fan.sys.StrBuf.prototype.m_capacity = null;

fan.sys.StrBuf.prototype.clear = function()
{
  this.m_str = "";
  return this;
}

fan.sys.StrBuf.prototype.get = function(i)
{
  if (i < 0) i = this.m_str.length+i;
  if (i < 0 || i >= this.m_str.length) throw fan.sys.IndexErr.make(i);
  return this.m_str.charCodeAt(i);
}

fan.sys.StrBuf.prototype.getRange = function(range)
{
  var size = this.m_str.length;
  var s = range.$start(size);
  var e = range.$end(size);
  if (e+1 < s) throw fan.sys.IndexErr.make(range);
  return this.m_str.substr(s, (e-s)+1);
}

fan.sys.StrBuf.prototype.set = function(i, ch)
{
  if (i < 0) i = this.m_str.length+i;
  if (i < 0 || i >= this.m_str.length) throw fan.sys.IndexErr.make(i);
  this.m_str = this.m_str.substr(0,i) + String.fromCharCode(ch) + this.m_str.substr(i+1);
  return this;
}

fan.sys.StrBuf.prototype.join = function(x, sep)
{
  if (sep === undefined) sep = " ";
  var s = (x == null) ? "null" : fan.sys.ObjUtil.toStr(x);
  if (this.m_str.length > 0) this.m_str += sep;
  this.m_str += s;
  return this;
}

fan.sys.StrBuf.prototype.insert = function(i, x)
{
  var s = (x == null) ? "null" : fan.sys.ObjUtil.toStr(x);
  if (i < 0) i = this.m_str.length+i;
  if (i < 0 || i > this.m_str.length) throw fan.sys.IndexErr.make(i);
  this.m_str = this.m_str.substr(0,i) + s + this.m_str.substr(i);
  return this;
}

fan.sys.StrBuf.prototype.remove = function(i)
{
  if (i < 0) i = this.m_str.length+i;
  if (i< 0 || i >= this.m_str.length) throw fan.sys.IndexErr.make(i);
  this.m_str = this.m_str.substr(0,i) + this.m_str.substr(i+1);
  return this;
}

fan.sys.StrBuf.prototype.removeRange = function(r)
{
  var s = r.$start(this.m_str.length);
  var e = r.$end(this.m_str.length);
  var n = e - s + 1;
  if (s < 0 || n < 0) throw fan.sys.IndexErr.make(r);
  this.m_str = this.m_str.substr(0,s) + this.m_str.substr(e+1);
  return this;
}

fan.sys.StrBuf.prototype.replaceRange = function(r, str)
{
  var s = r.$start(this.m_str.length);
  var e = r.$end(this.m_str.length);
  var n = e - s + 1;
  if (s < 0 || n < 0) throw fan.sys.IndexErr.make(r);
  this.m_str = this.m_str.substr(0,s) + str + this.m_str.substr(e+1);
  return this;
}

fan.sys.StrBuf.prototype.isEmpty = function()
{
  return this.m_str.length == 0;
}

fan.sys.StrBuf.prototype.size = function()
{
  return this.m_str.length;
}

fan.sys.StrBuf.prototype.toStr = function()
{
  return this.m_str;
}

fan.sys.StrBuf.prototype.out = function()
{
  return new fan.sys.StrBufOutStream(this);
}

//////////////////////////////////////////////////////////////////////////
// Static Methods
//////////////////////////////////////////////////////////////////////////

fan.sys.StrBuf.make = function() { return new fan.sys.StrBuf(); }